client/write_test.go

```diff
@@ -107,3 +112,104 @@ func setShardStates(t *testing.T, s *session, host topology.Host, state shard.St
 		hostShard.SetState(state)
 	}
 }
+
+type fakeHost struct{ id string }
+
+func (f fakeHost) ID() string      { return f.id }
+func (f fakeHost) Address() string { return "" }
+func (f fakeHost) String() string  { return "" }
+
+func TestErrorRetryability(t *testing.T) {
+	var wState *writeState
+	var s *session
+	var host topology.Host
+	var writeWg sync.WaitGroup
+
+	testSetup := func() {
+		ctrl := gomock.NewController(t)
+		defer ctrl.Finish()
+
+		s = newDefaultTestSession(t).(*session)
+		w := newWriteStub()
+
+		var completionFn completionFn
+		enqueueWg := mockHostQueues(ctrl, s, sessionTestReplicas, []testEnqueueFn{func(idx int, op op) {
+			completionFn = op.CompletionFn()
+		}})
+
+		require.NoError(t, s.Open())
+		defer func() {
+			require.NoError(t, s.Close())
+		}()
+
+		host = s.topoMap.Hosts()[0] // any host
+
+		wState = getWriteState(s)
+		wState.incRef() // for the test
+		wState.incRef() // allow introspection
+
+		// Begin write
+		writeWg.Add(1)
+		go func() {
+			s.Write(w.ns, w.id, w.t, w.value, w.unit, w.annotation)
+			writeWg.Done()
+		}()
+
+		// Callbacks
+
+		enqueueWg.Wait()
+		require.True(t, s.topoMap.Replicas() == sessionTestReplicas)
+		for i := 0; i < s.topoMap.Replicas(); i++ {
+			completionFn(host, nil) // maintain session state
+		}
+	}
+
+	testCleanup := func() {
+		wState.decRef() // end introspection
+		writeWg.Wait()  // wait for write to complete
+	}
```

Comment:
Instead of a testsetup and a testcleanup - perhaps instead use a test for each one of these cases for better isolating an issue if it comes up.  i.e. instead of one huge test failing, you have one of many smaller ones failing so you can track down the issue faster.

i.e.

```go
type testWriteFn func(
  	wState *writeState,
  	s *session,
  	host topology.Host,
  	writeWg sync.WaitGroup,
)

func testWrite(t *testing.T, fn testWriteFn) {
  	// all the code for test setup

  	fn(wState, s, host, writeWg)

  	// all the code for test cleanup
}

// now write tests 

func TestWriteNonRetryableError(t *testing.T) {
    testWrite(t,  func(wState *writeState, s *session, host topology.Host, writeWg sync.WaitGroup) {
 	 	wState.completionFn(host, xerrors.NewNonRetryableError(errors.New("")))
 	  	require.True(t, len(wState.errors) == 1)
 	  	assert.True(t, xerrors.IsNonRetryableError(wState.errors[0]))
    })
}

// other tests